File: psets/pset6/credit/credit.py
from cs50 import get_string
import math

# Prompts the user for the credit card number
card = get_string("Number: ")

# Calculates the number of digits
n = len(card)
csa = []

# Multiplies every other digit by 2
for i in range(n - 2, -1, -2):
    csa.append(int(card[i]) * 2)

# Calculates the sum of all the digits
for i in range(len(csa)):
    if int(csa[i]) >= 10:
        csa[i] = (csa[i] % 10) + (math.floor(csa[i] / 10))

a = sum(csa)

csb = []

# Calculates the sum of the rest of the digits
for i in range(n - 1, -1, -2):
    csb.append(int(card[i]))

b = sum(csb)

cs = a + b

# Checks if the checksu is valid
if cs % 10 == 0:
    # Checks for AMEX cards
    if n == 15:
        if card[0] == '3':
            if card[1] == 4 or card[1] == '7':
                print("AMEX")
            else:
                print("INVALID")
        else:
            print("INVALID")
    # Checks for VISA cards
    elif n == 13:
        if card[0] == '4':
            print("VISA")
        else:
            print("INVALID")
    # Checks for VISA or MASTERCARD cards
    elif n == 16:
        if card[0] == '4':
            print("VISA")
        elif card[0] == '5':
            if int(card[1]) >= 1 and int(card[1]) <= 5:
                print("MASTERCARD")
            else:
                print("INVALID")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")

File: psets/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long card = get_long("Number: ");     //prompts the user for the credit card number

    int n = 1;
    long i;

    for (i = 10; card > i; i *= 10)     //finds the number of digits
    {
        n++;
    }

    int a = 0;

    for (long j = 100; j <= i; j *= 100)     //calculates the first part of the checksum
    {
        a += (2 * ((card % j) / (j / 10)) / 10) + (2 * ((card % j) / (j / 10)) % 10);
    }

    int b = 0;

    for (long j = 10; j <= i; j *= 100)     //calculates the second part of the checksum
    {
        b += ((card % j) / (j / 10));
    }

    if ((a + b) % 10 == 0)     //checks if it's valid
    {
        if (n == 15)     //checks for AMEX
        {
            if (card / (i / 10) == 3)     //checks the first digit
            {
                if ((card % (i / 10)) / (i / 100) == 4 || (card % (i / 10)) / (i / 100) == 7)     //checks the second digit
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (n == 13)     //checks
        {
            if (card / (i / 10) == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (n == 16)
        {
            if (card / (i / 10) == 4)
            {
                printf("VISA\n");
            }
            else if (card / (i / 10) == 5)
            {
                if ((card % (i / 10)) / (i / 100) >= 1 && (card % (i / 10)) / (i / 100) <=5)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}